validator.go

```diff
@@ -0,0 +1,132 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package wrp
+
+import (
+	"errors"
+
+	"go.uber.org/multierr"
+)
+
+var (
+	ErrInvalidTypeValidator = errors.New("invalid WRP message type validator")
+	ErrInvalidMsgType       = errors.New("invalid WRP message type")
+)
+
+// AlwaysInvalid doesn't validate anything about the message and always returns an error.
+var AlwaysInvalid ValidatorFunc = func(m Message) error { return ErrInvalidMsgType }
+
+// AlwaysValid doesn't validate anything about the message and always returns nil.
+var AlwaysValid ValidatorFunc = func(msg Message) error { return nil }
+
+// Validator is a WRP validator that allows access to the Validate function.
+type Validator interface {
+	Validate(m Message) error
+}
+
+// Validators is a WRP validator that ensures messages are valid based on
+// message type and each validator in the list.
+type Validators []Validator
+
+// Validate runs messages through each validator in the validators list.
+// It returns as soon as the message is considered invalid, otherwise returns nil if valid.
+func (vs Validators) Validate(m Message) error {
+	var err error
+	for _, v := range vs {
+		err = multierr.Append(err, v.Validate(m))
+	}
+
+	return err
+}
+
+// ValidatorFunc is a WRP validator that takes messages and validates them
+// against functions.
+type ValidatorFunc func(Message) error
+
+// Validate executes its own ValidatorFunc receiver and returns the result.
+func (vf ValidatorFunc) Validate(m Message) error {
+	return vf(m)
+}
+
+// TypeValidator is a WRP validator that validates based on message type
+// or using the defaultValidators if message type is unfound.
+type TypeValidator struct {
+	m                 map[MessageType]Validator
+	defaultValidators Validators
+}
+
+// Validate validates messages based on message type or using the defaultValidators
+// if message type is unfound.
+func (m TypeValidator) Validate(msg Message) error {
+	vs := m.m[msg.MessageType()]
+	if vs == nil {
+		return m.defaultValidators.Validate(msg)
+	}
+
+	return vs.Validate(msg)
+}
+
+// NewTypeValidator is a TypeValidator factory.
+func NewTypeValidator(m map[MessageType]Validator, defaultValidators ...Validator) (TypeValidator, error) {
+	if m == nil {
+		return TypeValidator{}, ErrInvalidTypeValidator
+	}
+
+	for _, v := range m {
+		if err := validateValidator(v); err != nil {
+			return TypeValidator{}, err
+		}
+	}
+
+	if len(defaultValidators) == 0 {
+		defaultValidators = Validators{AlwaysInvalid}
+	}
+
+	for _, v := range defaultValidators {
+		if err := validateValidator(v); err != nil {
+			return TypeValidator{}, err
+		}
+	}
+
+	return TypeValidator{
+		m:                 m,
+		defaultValidators: defaultValidators,
+	}, nil
+}
+
+// validateValidator validates a given Validator.
+func validateValidator(v Validator) error {
+	switch vs := v.(type) {
+	case Validators:
```

Comment:
Rather than having special logic to run nil checks on `Validators`, I think we should have a nil check within `Validators.Validate()`.  If a validator in the `Validators` slice is nil, it should be skipped (but the rest run).  Then, we don't need this function at all (can do a simple nil check in the New function), and this nil case is covered when `Validators` is used on its own, outside of a `TypeValidator`.

Reply:
So nil validators will work the same as providing the `AlwaysValid` validator. I'm good with that.